_posts/2019-05-10-server-rendering-responsively.md

```diff
@@ -0,0 +1,189 @@
+(reluctantly) created our own user-agent detection library,
+[@artsy/detect-responsive-traits](https://github.com/artsy/detect-responsive-traits). This data is very specific to
+the browsers & devices we support on [artsy.net](artsy.net), but it's the data we need to optimize our responsive
+rendering. We're using it to determine if your browser is likely going to use only the mobile breakpoint of our
+app, in which case we don't have to also render the desktop version.
+
+We aren't doing any detection of desktop browsers. Maybe this is something we do in the future, but for now we are
+more concerned with mobile users getting less content sent over their 3G connection.
+
+## Why didn't you use \_\_\_?
+
+Those are our primary tools for combining SSR with responsive design! They work well for us. We considered many
+many other options along the way. Here are a couple:
+
+### [`react-media`](https://github.com/ReactTraining/react-media) or [`react-responsive`]()
+
+We investigated both `react-media` and `react-responsive`, but found that they didn't approach the SSR side of the
+problem as deeply as we needed.
+
+We also weren't fans of the imperative API in `react-media`. We started with a similar API when building
+`react-responsive-media`, but learned that it limited our ability to render all breakpoints from the server. The
+API requires only one branch to be rendered, and this caused the aforementioned mismatches when the app would
+hydrate.
+
+With `react-responsive`, we didn't like that it relied on user agent detection as its primary method of handling
+SSR.
+
+(TODO: are those statements correct?)
+
+### CSS
+
+We explored the idea of rendering all breakpoints from the server, and hiding the non-matching branches with CSS.
+The issue with this approach is that you have many components that are mounted & rendered unnecessarily. Aside from
+the performance hit you take for rendering components your user isn't seeing, even worse is the potential for
+duplicate side-effects.
+
+Imagine a component that, when rendered, emits a call to an analytics service. If this component exists in both a
+mobile and desktop branch, you're now double-stuffing your analytics. Hopefully your analytics service is smart
+enough to count both calls, but it's still a bad idea to duplicate components that have side-effects.
+
+## What's left to solve?
+
+Our SSR & responsive design toolbox does a lot of things well. We get great performance from both the server and
+client. Our site looks great on any device.
+
+We do have some SEO concerns, though. Since we're server-rendering multiple breakpoints, it's likely that search
+engine bots are seeing double the content on our pages. We _think_ this is okay.
+[Google WebMasters](https://youtu.be/WsgrSxCmMbM) says it's okay. We haven't noticed any awful side-effects from
+this yet, but SEO is a bit of a dark art, yeah?
```

Comment:
Could reenforce the point that once things mount on the client the duplication is swept, leaving behind a clean, proper page -- that (may) contribute to the lack of negative effects.

Reply:
I don't think I could state that with any confidence. Can you? (Not challenging, just asking. If we're all uncertain, I don't want to further muddy it.)

Reply:
The suggestion was more tangential / strategic -- it's an opportunity to state (once more) that once the user is ready to start interacting with the page there isn't markup for every breakpoint on the dom, and that fact may or may not contribute to our lack of SEO issues. Agree that it could muddy things a bit unless carefully worded. (No strong opinions about this here tho, just a thought.)